### PR TITLE
Podman Prototype and Unit test

### PR DESCRIPTION
**Pull request by luluz66**

### CODE REVIEW

File: enterprise/server/remote_execution/containers/podman/podman.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	"github.com/buildbuddy-io/buildbuddy/server/util/background"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/random"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+)
+
+var (
+	containerFinalizationTimeout = 10 * time.Second
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/containers/podman/podman.go:23]: nit: Can you comment this and how to set it?

File: enterprise/server/remote_execution/containers/podman/podman.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+)
+
+// podmanCommandContainer containerizes a command's execution using a Podman container.
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/containers/podman/podman.go:26]: love the go style comments :P If you added them to all public methods in this piece, I'd be so happy!

File: enterprise/server/remote_execution/containers/podman/podman.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		podmanRunArgs = append(podmanRunArgs, "--env", fmt.Sprintf("%s=%s", envVar.GetName(), envVar.GetValue()))
+	}
+	podmanRunArgs = append(podmanRunArgs, c.image)
+	podmanRunArgs = append(podmanRunArgs, command.Arguments...)
+	result = runPodman(ctx, "run", "", nil, nil, podmanRunArgs...)
+	if exitedCleanly := result.ExitCode >= 0; !exitedCleanly {
+		err = killContainerIfRunning(ctx, containerName)
+	}
+	if err != nil {
+		log.Printf("WARNING: Failed to shut down docker container: %s\n", err.Error())
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/containers/podman/podman.go:84]: log.Warningf

**Review by tylerwilliams — APPROVED** [review]
LGTM

This is dope. Couple nits, and I added Brandon to take a look too, but I think this is a great start!

File: enterprise/server/remote_execution/containers/podman/podman.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	result := &interfaces.CommandResult{
+		CommandDebugString: fmt.Sprintf("(podman) %s", command.GetArguments()),
+		ExitCode:           commandutil.NoExitCode,
+	}
+	containerName, err := generateContainerName()
+	if err != nil {
+		result.Error = status.UnavailableErrorf("failed to generate podman container name: %s", err)
+		return result
+	}
+	if err := c.PullImageIfNecessary(ctx); err != nil {
```
Comment (bduffany) [inline enterprise/server/remote_execution/containers/podman/podman.go:52]: See comments on `c.PullImageIfNecessary` below -- if you implement `PullImage` and `IsImageCached` then you can switch this to `container.PullImageIfNecessary(ctx, env, imageCacheAuthenticator, ...)`. The `PullImageIfNecessary` util func takes care of auth for locally cached images.

(Using that func would require plumbing the imageCacheAuthenticator through to `podmanCommandContainer`.)

File: enterprise/server/remote_execution/containers/podman/podman.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	// Try to avoid the `pull` command which results in a network roundtrip.
+	listResult := runPodman(ctx, "images", "", nil, nil, "--filter=reference="+r.image, "--format={{.ID}}")
+	if listResult.Error != nil {
+		return listResult.Error
+	}
+	if strings.TrimSpace(string(listResult.Stdout)) != "" {
+		// Found at least one image matching the ref; `docker run` should succeed
+		// without pulling the image.
+		return nil
+	}
```
Comment (bduffany) [inline enterprise/server/remote_execution/containers/podman/podman.go:137]: This logic could go in `IsImageCached`

File: enterprise/server/remote_execution/containers/podman/podman.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	pullResult := runPodman(ctx, "pull", "", nil, nil, r.image)
+	if pullResult.Error != nil {
+		return pullResult.Error
+	}
+	return nil
```
Comment (bduffany) [inline enterprise/server/remote_execution/containers/podman/podman.go:142]: This logic could go in `PullImage`

**Review by bduffany — APPROVED** [review]
Looking good so far :+1: I left some notes in case you want to refactor the image pull logic to handle private registry auth sooner rather than later, but it's also cool to do it in a separate PR

File: enterprise/server/remote_execution/containers/podman/podman.go (hunk 6 above)
Comment (luluz66) [reply enterprise/server/remote_execution/containers/podman/podman.go]: Done

File: enterprise/server/remote_execution/containers/podman/podman.go (hunk 5 above)
Comment (luluz66) [reply enterprise/server/remote_execution/containers/podman/podman.go]: Done

File: enterprise/server/remote_execution/containers/podman/podman.go (hunk 4 above)
Comment (luluz66) [reply enterprise/server/remote_execution/containers/podman/podman.go:53]: I will plumbing imageCacheAuthenticator in a later PR.

File: enterprise/server/remote_execution/containers/podman/podman.go (hunk 1 above)
Comment (luluz66) [reply enterprise/server/remote_execution/containers/podman/podman.go:24]: Done

File: enterprise/server/remote_execution/containers/podman/podman.go (hunk 2 above)
Comment (luluz66) [reply enterprise/server/remote_execution/containers/podman/podman.go:27]: The interface commanderContainer has all the comments for the public methods.

**Pull request merged**